code/assignTime.m

```diff
@@ -0,0 +1,279 @@
+% flag as gap
+indices_systemTickFlagged = find (abs(expectedElapsed(2:end) - diff_systemTick(2:end)) > 0.1*expectedElapsed(2:end));
+
+% All packets flagged as end of continuous chunks
+allFlaggedIndices = unique([indices_changeFs; indices_timestampFlagged;...
+    indices_dataTypeSequenceFlagged; indices_systemTickFlagged],'sorted') ;
+
+%%
+disp('Chunking data')
+% Determine indices of packets which correspond to each data chunk
+if ~isempty(allFlaggedIndices)
+    counter = 1;
+    chunkIndices = cell(1,length(allFlaggedIndices) + 1);
+    for iChunk = 1:length(allFlaggedIndices)
+        if iChunk == 1
+            chunkIndices{counter} = (1:allFlaggedIndices(1));
+            currentStartIndex = 1;
+            counter = counter + 1;
+        elseif iChunk == length(allFlaggedIndices)
+            chunkIndices{counter} = allFlaggedIndices(currentStartIndex) + 1:allFlaggedIndices(currentStartIndex + 1);
+            chunkIndices{counter + 1} = allFlaggedIndices(currentStartIndex + 1) + 1:numPackets;
+        else
+            chunkIndices{counter} = allFlaggedIndices(currentStartIndex) + 1:allFlaggedIndices(currentStartIndex + 1);
+            currentStartIndex = currentStartIndex + 1;
+            counter = counter + 1;
+        end
+     end
+else
+    % No identified missing packets, all packets in one chunk
+    chunkIndices{1} = 1:numPackets;
+end
+
+%%
+% Loop through each chunk to determine offset to apply (as determined by
+% average difference between packetGenTime and expectedElapsed)
+disp('Determining start time of each chunk')
+
+% PacketGenTime in ms; convert difference to 1e-4 seconds, units of
+% systemTick and expectedElapsed
+diff_PacketGenTime = [1; diff(dataTable.PacketGenTime) * 1e1];
+
+numChunks = length(chunkIndices);
+chunksToExclude = [];
+meanError = NaN(1,numChunks);
+for iChunk = 1:numChunks
+    currentTimestampIndices = chunkIndices{iChunk};
+    
+    % Chunks must have at least 2 packets in order to have a valid
+    % diff_systemTick -- thus if chunk only one packet, it must be excluded
+    if length(currentTimestampIndices) == 1
+        chunksToExclude = [chunksToExclude iChunk];
+    end
+    % Always exclude the first packet of the chunk, because don't have an
+    % accurate diff_systemTick value for this first packet
+    currentTimestampIndices = currentTimestampIndices(2:end);
+    
+    % Differences between adjacent PacketGenTimes (in units of 1e-4
+    % seconds)
+    error = expectedElapsed(currentTimestampIndices) - diff_PacketGenTime(currentTimestampIndices);
+    meanError(iChunk) = median(error);
+end
+%%
+% Create corrected timing for each chunk
+counter = 1;
+for iChunk = 1:numChunks
+    if ~ismember(iChunk,chunksToExclude)
+        alignTime = dataTable.PacketGenTime(chunkIndices{iChunk}(1));
+        % alignTime in ms; meanError in units of systemTick
+        correctedAlignTime(counter) = alignTime + meanError(iChunk)*1e-1;
+        % TO DO: add or subtract meanError above??
```

Comment:
Adding or subtracting depends on the assumption. If you believe the packet represents the correct amount of elapsed time, then adding should be the operation. This is because you calculated the median difference between the expected elapsed time according to the packet size and the packet generation time. The number of time units will be negative if the packet gen time is more consistently later than the expected elapsed time, so adding is subtracting the bias. If you think the packet gen time is a more accurate representation of time, then you would want to subtract the value in meanError. 

If anyone has an opposing view, or I got my logic mixed up, please respond!

Reply:
This makes sense to me. I've added a long comment explaining the rationale.